### ISSUE TITLE
improve handling of search terms containing Lucene special characters

### ISSUE THREAD

**Issue by saschaszott** [DSpace/DSpace#9026]
The search query handling in DSpace 7.6 does not respect **Lucene special characters**.

I've submitted a publication item with title _Is this a test?_ in the new demo site

![image](https://github.com/DSpace/DSpace/assets/952735/029ee264-0739-4178-8b85-518b4d1f4b97)

(current URL: https://preview.dspace.org/entities/publication/3d92e8b1-8ad5-4373-90d2-a530037983dd)

The search query _Is this a test?_ does not retrieve this item.

The same result is obtained with the search query _test?_.

The current list of Lucene special characters is

```
+ - && || ! ( ) { } [ ] ^ " ~ * ? : \
```

To escape a special character a `\` before the character is needed.  SolrJ supports escaping of query terms: https://solr.apache.org/docs/8_11_0//solr-solrj/org/apache/solr/client/solrj/util/ClientUtils.html#escapeQueryChars-java.lang.String-

**Comment by tdonohue** [comment]
Pulling this over to the 7.6.x board alongside #8962 as they do seem related (and might be solved by a single PR).  A volunteer is needed to investigate ways to better support special characters in searches.

**Comment by aroman-arvo** [comment]
Escaping special characters can be problematic since their use is offered in the documentation "DSpace supports all Solr search syntax options, as all searches in DSpace are sent directly to Solr. For more examples, see the Solr documentation for the "Specifying Terms for the Standard Query Parser" (https://wiki.lyrasis.org/display/DSDOC7x/Search+-+Advanced)

We could be adjusting which characters are escaped and which are not and cases would continue to appear in which it does not work as expected.
I think that this search should show more results that needed, and that the user can filter it down with facets or adding more search terms. So the best solution that I can think that supports both functions, both the escaped option and the syntax with wildcards, is to transform the requests before launching them into the sun by concatenating both searches. That example should be:

`Is this a test? --> (Is this a test\?) OR (Is this a test?)`

**Comment by tdonohue** [comment]
@aroman-arvo : Good point.  Maybe this entire ticket is just a documentation issue?

It could be that we simply need to document which characters Lucene (and therefore DSpace) considers "special characters" and then document how to escape them in your searches.  That'd also give users more power, but they'd need to know that a search for `Is this a test?` should instead be written as `Is this a test\?` if they want to search on a literal question mark (`?`).

If that's true, I'd lean towards not changing the current behavior.  Instead, we could update the documentation at https://wiki.lyrasis.org/display/DSDOC7x/Search+-+Advanced to clearly note which characters are special characters and how to properly escape them in queries.

Do others agree?  Should we simply improve our documentation and close this ticket?

**Comment by mwoodiupui** [comment]
I agree:  if DSpace is to permit the use of Lucene query syntax in DSpace searches, then it is up to the user to indicate whether this is being done, and DSpace's responsibility is to document how to do or not do that as the user may require.  There is no way for DSpace to know which is wanted, in any given position in the search string, but that the user tell it.

**Comment by aroman-arvo** [comment]
@tdonohue 
I would vote for avoiding the user, especially the casual user, to have to master the operation of the search engine. 
Transforming the query into one that returns the aggregated results of both interpretations of the special characters is simple to implement and does not generate too much noise.

**Comment by tdonohue** [comment]
@aroman-arvo : I don't agree with modifying a user's query for the following reasons:
* I believe this will impact the relevance ranking from Solr, and users will have less control over relevance via advanced search techniques described at https://wiki.lyrasis.org/display/DSDOC7x/Search+-+Advanced  (In other words, matching Items may be returned in an order that looks semi-random... the best matches may no longer be at the top of the list.)
* I worry this will generate unnecessary "noise" as it will be more complex to track [search statistics](https://wiki.lyrasis.org/display/DSDOC7x/SOLR+Statistics#SOLRStatistics-SearchQueryStatistics) (which don't exist yet in 7.x but should at some point). In other words, it will be difficult to determine the original search text unless we somehow store/log it separate from the automatically modified search text.
* It also may impact user interface display. The user will expect to see their original search text in the search results page. But if we modify that search text, they will likely see the *modified search* in the user interface, as we have no way to store "original search text" separate from "modified search text".
* It gives advanced researchers/searchers (like librarians) _less control_, as it will send them results which may not match their original search query.  This means bug tickets like #8962 will occur more frequently, as "exact matches" will be nearly impossible to achieve.

Overall, I fear this is not as simple as it may seem, and I don't want DSpace to get in the habit of building custom search logic/algorithms (as that can get very complex very quickly).

I feel we have two primary options here:
1. Leave behavior as-is and document it better (as I initially [suggested above](https://github.com/DSpace/DSpace/issues/9026#issuecomment-1749176609))
2. Or, leave behavior as-is and add a "Did you mean?" style of response for queries that have special characters.  In other words, if the user searches on `Is this a test?` we could send them back initial response of "No matches" while also adding a message in the UI that says something like: 
    > Your query contained special characters and may not have matched all relevant objects in the system.  Would you like to [run query again with special characters escaped]?" 
    * The above text could then be hyperlinked to run the same query but with all known special characters automatically escaped.

**Comment by mwoodiupui** [comment]
Yes, and there are two groups of users with different expectations.  One expects a search input to be sought literally; the other wants the input to be a Lucene expression (or something of roughly equivalent power).  The two groups will never be reconciled.  So why not a simple checkbox to indicate "search literally" (wording subject to discussion).  If you uncheck the box then DSpace will send the query as-is; otherwise DSpace will escape every character that Lucene treats specially.  You get what you asked for, and DSpace does not have to guess what you meant.
